eras/shelley/impl/src/Cardano/Ledger/Shelley/AdaPots.hs

```diff
@@ -72,6 +92,85 @@ totalAdaES cs =
+        -- keyDepositAdaPot,  -- We don't count these two, as their
+        -- poolDepositAdaPot, -- sum is always depositsAdaPot
         depositsAdaPot,
         feesAdaPot
       } = totalAdaPotsES cs
+
+-- =============================================
+-- Produced and Consumed are specialized AdaPots
+-- relative to the actions of a TxBody
+
+-- | Itemizing what is consumed by a transaction
+data Consumed = Consumed
+  {conInputs :: !Coin, conRefunds :: !Coin, conWithdrawals :: !Coin}
+
+instance Show Consumed where
+  show (Consumed (Coin i) (Coin r) (Coin w)) =
+    "Consumed(Inputs " ++ show i ++ ", Refunds " ++ show r ++ ", Withdrawals " ++ show w ++ ") = " ++ show (i + r + w)
+
+-- | Itemizing what is Produced by a transaction
+data Produced = Produced
+  {proOutputs :: !Coin, proFees :: !Coin, proDeposits :: !Coin}
+
+instance Show Produced where
+  show (Produced (Coin out) (Coin f) (Coin d)) =
+    "Produced(Outputs " ++ show out ++ ", Fees " ++ show f ++ " Deposits " ++ show d ++ ") = " ++ show (out + f + d)
+
+-- =========================
+
+-- | Compute the Coin part of what is consumed by a TxBody, itemized as a 'Consume'
+consumedTxBody ::
+  ( HasField "_keyDeposit" pp Coin,
+    ShelleyEraTxBody era
+  ) =>
+  TxBody era ->
+  pp ->
+  DPState (EraCrypto era) ->
+  UTxO era ->
+  Consumed
+consumedTxBody txBody pp dpstate (UTxO u) = Consumed {conInputs = i, conRefunds = r, conWithdrawals = w}
+  where
+    i = coinBalance (UTxO (Map.restrictKeys u (txBody ^. inputsTxBodyL)))
+    r = keyTxRefunds pp dpstate txBody
+    w = fold . unWdrl $ txBody ^. wdrlsTxBodyL
+
+-- | Compute the Coin part of what is produced by a TxBody, itemized as a 'Produced'
+producedTxBody ::
+  ( ShelleyEraTxBody era,
+    HasField "_keyDeposit" pp Coin,
+    HasField "_poolDeposit" pp Coin
+  ) =>
+  TxBody era ->
+  pp ->
+  DPState (EraCrypto era) ->
+  Produced
+producedTxBody txBody pp dpstate = Produced {proOutputs = out, proFees = f, proDeposits = d}
+  where
+    out = coinBalance (txouts txBody)
+    f = txBody ^. feeTxBodyL
+    d = totalTxDeposits pp dpstate txBody
+
+-- ==============================================
+-- Compare two AdaPots, item by item
+
+pad :: Int -> String -> String
+pad n x = x ++ replicate (n - length x) ' '
+
+compareAdaPots :: String -> AdaPots -> String -> AdaPots -> String
+compareAdaPots xlabel x ylabel y =
+  unlines
+    [ pad n "field" ++ pad n xlabel ++ pad n ylabel ++ pad n "difference",
+      oneline "treasuryAdaPot" treasuryAdaPot,
+      oneline "reservesAdaPot" reservesAdaPot,
+      oneline "rewardsAdaPot" rewardsAdaPot,
+      oneline "utxoAdaPot" utxoAdaPot,
+      oneline "keyDepositAdaPot" keyDepositAdaPot,
+      oneline "poolDepositAdaPot" poolDepositAdaPot,
+      oneline "depositsAdaPot" depositsAdaPot,
+      oneline "feesAdaPot" feesAdaPot
+    ]
+  where
+    n = 25
+    oneline name f = pad n name ++ pad n (show (f x)) ++ pad n (show (f y)) ++ pad n (show (f y <-> f x))
```

Comment:
I'll post here one more time. This should go into `Test.cardano.Ledger.Shelley.Rules.TestChain`, because that is the only place where it is being used

Reply:
NO,  the problem is that you never debug test failures. This is the most useful tool ever to debug test failures, since it itemizes where things have changed. I use it in debugging by putting trace calls which use this function. In order for me to make such trace calls, this function must be in scope. If it is in the tests, it is not in scope in the src files. 

Once the the failures have been debugged, the trace calls are removed, and the it APPEARS that this has only one use. But in fact I have used it many times. It needs to be in scope. It is such a little thing, but it is the exception that breaks the rule.

Reply:
So now this has been moved to Cardano.Ledger.Shelley.Internal(compareAdaPots)

Reply:
Could we keep this part (slightly edited) in the haddock for the function :slightly_smiling_face: ?
> This is the most useful tool ever to debug test failures, since it itemizes where things have changed. Use it for debugging by putting trace calls composed with this function.